File: Template/Floyd.cpp
//? 플로이드
#include <cstdio>
#include <algorithm>
using namespace std;
const int INF = 1000000000; // 절대 나올 수 없는 경로값
 
int main(){
    int N, M, dist[100][100];
    scanf("%d %d", &N, &M);
    // dist 배열 초기화
    for(int i=0; i<N; i++)
        for(int j=0; j<N; j++)
            dist[i][j] = i==j ? 0 : INF;
    // 간선 정보 입력받음
    for(int i=0; i<M; i++){
        int a, b, c;
        scanf("%d %d %d", &a, &b, &c);
        dist[a-1][b-1] = min(dist[a-1][b-1], c);
    
    // 플로이드 와샬 알고리즘 적용
    for(int k=0; k<N; k++)
        for(int i=0; i<N; i++)
            for(int j=0; j<N; j++)
                dist[i][j] = min(dist[i][j], dist[i][k]+dist[k][j]);
    
    // 이제 dist 배열은 실제 최단 경로를 담고 있음
    for(int i=0; i<N; i++){
        for(int j=0; j<N; j++)
            printf("%d ", dist[i][j]);
        puts("");
    }
}

//? 방문 가능 체크
// https://www.acmicpc.net/problem/6156 
#include <algorithm>
#include <iostream>
#include <limits>
#include <string>
#include <vector>

using std::cin; using std::cout;
using ull = unsigned long long;
using pii = std::pair<int, int>;
const int INF = std::numeric_limits<int>::max();

bool arr[101][101];

int main() {
  std::ios_base::sync_with_stdio(false);
  cin.tie(NULL); cout.tie(NULL);

  int n, m; cin >> n >> m;

  for (int i = 0; i < m; ++i) {
    int u, v; cin >> u >> v;
    arr[u][v] = true;
  }

  for (int k = 1; k <= n; ++k) {
    for (int i = 1; i <= n; ++i) {
      for (int j = 1; j <= n; ++j) {
        if (arr[i][k] & arr[k][j]) {
          arr[i][j] = true;
        }
      }
    }
  }

  int answer = 0;
  for (int i = 1; i <= n; ++i) {
    int count = 0;
    for (int j = 1; j <= n; ++j) {
      if (arr[i][j] || arr[j][i]) {
        count++;
      }
    }
    if (count == n-1) {
      answer++;
    }
  }

  cout << answer << '\n';
}
// g++ -o a -std=c++17 *.cpp

File: Template/Python/Floyd.py
# https://www.acmicpc.net/problem/11780
import sys

INF = 100_000_000
dp = [[0 if i == j else INF for i in range(101)] for j in range(101)]
pathInfo = [[-1 for _ in range(101)] for _ in range(101)]
path = []

def find(s, e):
  if pathInfo[s][e] != -1:
    find(s, pathInfo[s][e])
    path.append(pathInfo[s][e])
    find(pathInfo[s][e], e)

n = int(sys.stdin.readline())
m = int(sys.stdin.readline())

for _ in range(m):
  u, v, w = map(int, sys.stdin.readline().split())
  dp[u][v] = min(dp[u][v], w)

for k in range(1, n+1):
  for i in range(1, n+1):
    for j in range(1, n+1):
      cost = dp[i][k] + dp[k][j];
      if dp[i][j] > cost:
        dp[i][j] = cost
        pathInfo[i][j] = k

for i in range(1, n+1):
  for j in range(1, n+1):
    print(dp[i][j] if dp[i][j] != INF else 0, end=' ')
  print()

for i in range(1, n+1):
  for j in range(1, n+1):
    path.clear()
    if (i == j or dp[i][j] == INF):
      print(0, end=' ')
    else:
      path.append(i)
      find(i, j)
      path.append(j)
      print(len(path), end=' ')
      for p in path:
        print(p, end=' ')
    print()
